drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,410 @@
+use crate::ServerImpl;
+use drv_sidecar_front_io::transceivers::{
+    FpgaController, FpgaPortMasks, PortLocation,
+};
+use hubpack::SerializedSize;
+use task_net_api::*;
+use transceiver_messages::{
+    message::*,
+    mgmt::{MemoryRegion, UpperPage},
+    Error, ModuleId,
+};
+use userlib::hl::sleep_for;
+
+/// Convert from the over-the-network type to our local port mask type
+fn get_mask(m: ModuleId) -> Result<FpgaPortMasks, Error> {
+    let fpga_ports: u16 = m.ports.0;
+    match m.fpga_id {
+        0 => Ok(FpgaPortMasks {
+            left: fpga_ports,
+            right: 0,
+        }),
+        1 => Ok(FpgaPortMasks {
+            left: 0,
+            right: fpga_ports,
+        }),
+        i => Err(Error::InvalidFpga(i)),
+    }
+}
+
+impl ServerImpl {
+    /// Attempt to read and handle data from the `net` socket
+    pub fn check_net(
+        &mut self,
+        rx_data_buf: &mut [u8],
+        tx_data_buf: &mut [u8],
+    ) {
+        const SOCKET: SocketName = SocketName::transceivers;
+
+        match self.net.recv_packet(
+            SOCKET,
+            LargePayloadBehavior::Discard,
+            rx_data_buf,
+        ) {
+            Ok(mut meta) => {
+                // Modify meta.size based on the output packet size
+                meta.size = match hubpack::deserialize(rx_data_buf) {
+                    Ok((msg, data)) => {
+                        let (reply, data_len) = match self.handle_message(
+                            msg,
+                            data,
+                            &mut tx_data_buf[Message::MAX_SIZE..],
+                        ) {
+                            Ok(r) => r,
+                            Err(e) => (HostResponse::Error(e), 0),
+                        };
+                        let out = Message {
+                            header: msg.header,
+                            modules: msg.modules,
+                            body: MessageBody::HostResponse(reply),
+                        };
+                        // Serialize into the front of the tx buffer
+                        let msg_len =
+                            hubpack::serialize(tx_data_buf, &out).unwrap();
+
+                        // At this point, any supplementary data is written to
+                        // tx_buf[Message::MAX_SIZE..].  Let's shift it
+                        // backwards based on the size of the leading `Message`:
+                        tx_data_buf.copy_within(
+                            Message::MAX_SIZE..(Message::MAX_SIZE + data_len),
+                            msg_len,
+                        );
+                        (msg_len + data_len) as u32
+                    }
+                    Err(_e) => hubpack::serialize(
+                        tx_data_buf,
+                        &Message {
+                            header: Header {
+                                version: 1,
+                                message_id: u64::MAX,
```

Comment:
Yeah, this branch is only taken when deserialization fails (which shouldn't happen, considering UDP has checksums, but 🤷🏻 ).

I think it makes sense to try deserializing just the `Header`, since we should be able to deserialize that independently of the rest of the message; if that fails _too_, any opinions on what to do?

(We could certainly drop it, rather than replying at all, but if we want to report something to the host, we'd have to pick a message ID with nothing to go on)

Reply:
Agreed it's all pretty unlikely!

I also think that it makes sense to drop the packet if the `Header` can't be deserialized. The host will currently keep retrying until it gets a positive ACK from the SP, so dropping it on the SP-side does not affect that behavior. The alternative, such as sending a message ID of `u64::MAX`, seems no better. Even _if_ the host could match that up with its outstanding request, it would just retry it anyway.

Reply:
> Yeah, this branch is only taken when deserialization fails (which shouldn't happen, considering UDP has checksums, but 🤷🏻 ).

I've never seen a "packet was corrupt" deserialization failure, but in development I see deserialization failures _all the time_ when I've added or changed message types and only updated or rebuilt one side. Splitting the deserialization of the header handles all of those cases cleanly, since the header never changes.

FWIW in MGS I went with "reply with a bogus message ID" if deserializing the header fails; MGS can't actually do anything with that to pair it up with the request, so it'll eventually resend its request anyway, but again in development in particular it's nice to be able to tell the difference between "I'm not getting the response I expect, but I am seeing packets with bogus message IDs" from "the SP isn't responding".

Reply:
I think this is now changed to match [the host behavior](https://github.com/oxidecomputer/transceiver-control/pull/13#discussion_r1024496259), but double-checking would be appreciated!

Reply:
GH is so hard for me to navigate, that I can't actually find this comment thread relative to the source. It looks like this is handled by returning an error message body from `handle_host_request` and the response `out` in `net_api.rs:192` always uses the incoming message header. If that's true, then I agree it looks good!